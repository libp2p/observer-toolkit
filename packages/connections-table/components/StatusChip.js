import React from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { statusNames } from '@libp2p-observer/data'
import { Icon } from '@libp2p-observer/sdk'

const Container = styled.span`
  display: inline-block;
  padding: ${({ theme }) => theme.spacing(0.5)};
  background: ${({ theme, getColor }) => getColor(theme, 0.3)};
  color: ${({ theme, getColor }) => getColor(theme, 1)};
  white-space: nowrap;
  font-size: 8pt;
  font-weight: 600;
`

const ChipText = styled.span`
  line-height: 1em;
`

function StatusChip({ status }) {
  const chipMap = {
    ACTIVE: {
      icon: 'check',
      getColor: (theme, opacity) => theme.color('tertiary', 2, opacity),
    },
    OPENING: {
      icon: 'opening',
      getColor: (theme, opacity) => theme.color('tertiary', 1, opacity),
    },
    CLOSING: {
      icon: 'closing',
      getColor: (theme, opacity) => theme.color('highlight', 0, opacity),
    },
    CLOSED: {
      icon: 'closed',
      getColor: (theme, opacity) => theme.color('highlight', 1, opacity),
    },
    ERROR: {
      icon: 'cancel',
      getColor: (theme, opacity) => theme.color('contrast', 2, opacity),
    },
  }

  if (!chipMap[status])
    throw new Error(`Status name "${status}" not recognised`)
  const { getColor, icon } = chipMap[status]

  return (
    <Container getColor={getColor}>
      <Icon type={icon} />
      <ChipText>{status}</ChipText>
    </Container>
  )
}

StatusChip.propTypes = {
  status: T.oneOf(Object.values(statusNames)),
}

export default StatusChip
