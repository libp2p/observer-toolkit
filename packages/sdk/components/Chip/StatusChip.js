import React, { useContext } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { statusNames } from '@libp2p/observer-data'
import { StaticChip, ConfigContext, Tooltip } from '@libp2p/observer-sdk'

const TooltipContainer = styled.div`
  width: 200%;
`

const options = {
  ACTIVE: {
    icon: 'check',
    colorKey: 'tertiary',
    colorIndex: 2,
  },
  OPENING: {
    icon: 'opening',
    colorKey: 'tertiary',
    colorIndex: 1,
  },
  CLOSING: {
    icon: 'closing',
    colorKey: 'highlight',
    colorIndex: 0,
  },
  CLOSED: {
    icon: 'closed',
    colorKey: 'highlight',
    colorIndex: 1,
  },
  ERROR: {
    icon: 'cancel',
    colorKey: 'contrast',
    colorIndex: 1,
  },
}

// Don't fade so much it becomes unreadable
const minFadeOpacity = 0.2
function getFadeOpacity(value, maxValue) {
  if (typeof value !== 'number' || typeof maxValue !== 'number') return 1
  const fadeAmount = Math.max(0, 1 - value / maxValue)
  const fadeOpacity = minFadeOpacity + (1 - minFadeOpacity) * fadeAmount
  return fadeOpacity
}

const maxGlowOpacity = 0.4
function getGlowOpacity(value, maxValue) {
  if (typeof value !== 'number' || typeof maxValue !== 'number') return 0
  const glowAmount = Math.max(0, 1 - value / maxValue)
  const glowOpacity = glowAmount * maxGlowOpacity
  return glowOpacity
}

function StatusChip({ status, timeOpen, timeClosed, duration }) {
  const config = useContext(ConfigContext)
  const expiryMs = config ? config.getRetentionPeriodMs() : 0
  const fade = status === 'CLOSED' ? getFadeOpacity(timeClosed, expiryMs) : null
  const glow =
    status === 'ACTIVE' ? getGlowOpacity(timeOpen, duration * 4) : null

  const secondsUntilExpire = fade && Math.round((expiryMs - timeClosed) / 1000)

  const tooltipText =
    (glow && `New connection, open ${timeOpen} ms`) ||
    (fade &&
      `Fading off table in ${secondsUntilExpire} second${
        secondsUntilExpire === 1 ? '' : 's'
      }`) ||
    null

  return (
    <Tooltip
      content={tooltipText}
      side={'right'}
      toleranceY={null}
      fixOn={'never'}
      override={{ Content: TooltipContainer }}
    >
      <StaticChip type={status} options={options} fade={fade} glow={glow}>
        {status}
      </StaticChip>
    </Tooltip>
  )
}

StatusChip.propTypes = {
  status: T.oneOf(Object.values(statusNames)),
  timeOpen: T.number,
  timeClosed: T.number,
  duration: T.number,
}

export default StatusChip
