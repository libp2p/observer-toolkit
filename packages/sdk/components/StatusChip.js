import React, { useContext } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { statusNames } from '@libp2p-observer/data'
import { Chip, RuntimeContext, Tooltip } from '@libp2p-observer/sdk'

const TooltipContainer = styled.div`
  width: 200%;
`

const options = {
  ACTIVE: {
    icon: 'check',
    colorKey: 'tertiary',
    colorIndex: 2,
  },
  OPENING: {
    icon: 'opening',
    colorKey: 'tertiary',
    colorIndex: 1,
  },
  CLOSING: {
    icon: 'closing',
    colorKey: 'highlight',
    colorIndex: 0,
  },
  CLOSED: {
    icon: 'closed',
    colorKey: 'highlight',
    colorIndex: 1,
  },
  ERROR: {
    icon: 'cancel',
    colorKey: 'contrast',
    colorIndex: 1,
  },
}

// Don't fade so much it becomes unreadable
const minFadeOpacity = 0.2
function getFadeOpacity(value, maxValue) {
  const fadeAmount = Math.max(0, 1 - value / maxValue)
  const fadeOpacity = minFadeOpacity + (1 - minFadeOpacity) * fadeAmount
  return fadeOpacity
}

const maxGlowOpacity = 0.4
function getGlowOpacity(value, maxValue) {
  const glowAmount = Math.max(0, 1 - value / maxValue)
  const glowOpacity = glowAmount / maxGlowOpacity - (1 - maxGlowOpacity)
  return Math.max(glowOpacity, 0)
}

function StatusChip({ status, timeOpen, timeClosed, duration }) {
  const runtime = useContext(RuntimeContext)
  const expiryMs = runtime ? runtime.getKeepStaleDataMs() : 0
  const fade = status === 'CLOSED' ? getFadeOpacity(timeClosed, expiryMs) : null
  const glow = status === 'ACTIVE' ? getGlowOpacity(timeOpen, duration * 2) : null

  const secondsUntilExpire = fade && Math.round((expiryMs - timeClosed) / 1000)

  const tooltipText =
    (glow && `New connection, open ${timeOpen} ms`) ||
    (fade &&
      `Fading off table in ${secondsUntilExpire} second${
        secondsUntilExpire === 1 ? '' : 's'
      }`) ||
    null

  return (
    <Tooltip
      content={tooltipText}
      side={'right'}
      toleranceY={null}
      fixOn={'never'}
      override={{ Content: TooltipContainer }}
    >
      <Chip type={status} options={options} fade={fade} glow={glow}>
        {status}
      </Chip>
    </Tooltip>
  )
}

StatusChip.propTypes = {
  status: T.oneOf(Object.values(statusNames)),
}

export default StatusChip
