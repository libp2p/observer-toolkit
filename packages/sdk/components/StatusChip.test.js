import React from 'react'
import { catchErrorSilently, renderWithTheme } from '@libp2p/observer-testing'
import StatusChip from './StatusChip'

describe('status chip', () => {
  it('throws if trying to render an invalid status', () => {
    expect(
      catchErrorSilently(() =>
        renderWithTheme(<StatusChip status="invalid status" />)
      )
    ).toBeInstanceOf(Error)
  })

  it('renders ACTIVE status as expected', () => {
    const { asFragment } = renderWithTheme(<StatusChip status="ACTIVE" />)
    expect(asFragment()).toMatchSnapshot()
  })
})
