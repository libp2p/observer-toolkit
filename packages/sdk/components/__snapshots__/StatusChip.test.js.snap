// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`status chip renders ACTIVE status as expected 1`] = `
<DocumentFragment>
  .c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
}

.c1 {
  display: inline-block;
  padding: 0px 4px;
  background: rgba(13,62,80,0.15);
  color: rgba(13,62,80,1);
  white-space: nowrap;
  font-size: 8pt;
  font-weight: 600;
}

.c3 {
  line-height: 2em;
  display: inline-block;
  padding: 4px;
  vertical-align: middle;
}

<div
    class="c0"
  >
    <span
      class="c1"
    >
      <span
        class="c2"
        role="img"
      >
        <svg
          viewBox="0 0 96 96"
        >
          <g
            data-name="Layer 34"
          >
            <path
              d="M48 19a29 29 0 11-29 29 29 29 0 0129-29m0-4a33 33 0 1033 33 33 33 0 00-33-33z"
              fill="currentColor"
            />
            <path
              d="M35.32 46.81l8.73 11.46 16.63-20.54"
              fill="none"
              stroke="currentColor"
              stroke-miterlimit="10"
              stroke-width="4"
            />
          </g>
        </svg>
      </span>
      <span
        class="c3"
      >
        ACTIVE
      </span>
    </span>
  </div>
</DocumentFragment>
`;
